academic_observatory_workflows/workflows/pubmed_telescope.py

```diff
@@ -0,0 +1,1623 @@
+        """
+
+        self.filename = filename
+        self.file_index = file_index
+        self.baseline = baseline
+        self.path_on_ftp = path_on_ftp
+        self.datafile_date = datafile_date
+        self.datafile_release: DatafileRelease = datafile_release
+
+        self.file_type = "jsonl.gz"
+
+    def __eq__(self, other):
+        if isinstance(other, Datafile):
+            return (
+                self.filename == other.filename
+                and self.file_index == other.file_index
+                and self.baseline == other.baseline
+                and self.path_on_ftp == other.path_on_ftp
+                and self.datafile_date == other.datafile_date
+                and self.datafile_release == other.datafile_release
+            )
+        return False
+
+    @staticmethod
+    def from_dict(dict_: Dict) -> Datafile:
+        filename = dict_["filename"]
+        file_index = dict_["file_index"]
+        baseline = dict_["baseline"]
+        path_on_ftp = dict_["path_on_ftp"]
+        datafile_date = pendulum.parse(dict_["datafile_date"])
+        datafile_release = dict_["datafile_release"]
+
+        return Datafile(
+            filename=filename,
+            file_index=file_index,
+            baseline=baseline,
+            path_on_ftp=path_on_ftp,
+            datafile_date=datafile_date,
+            datafile_release=datafile_release,
+        )
+
+    def to_dict(self) -> Dict:
+        return dict(
+            filename=self.filename,
+            file_index=self.file_index,
+            baseline=self.baseline,
+            path_on_ftp=self.path_on_ftp,
+            datafile_date=self.datafile_date.isoformat(),
+            datafile_release=self.datafile_release,
+        )
+
+    @property
+    def download_file_path(self):
+        assert self.datafile_release is not None, "Datafile.download_folder: self.datafile_release is None"
+        return os.path.join(self.datafile_release.download_folder, self.filename)
+
+    @property
+    def transform_path(self):
+        assert self.datafile_release is not None, "Datafile.transform_path: self.datafile_release is None"
+        return self.datafile_release.transform_folder
+
+    def transform_file_path(self, entity_type: str) -> str:
+        """
+        Give path to the transform file depending on entity type.
+
+        :param entity_type: Type of the record, either "upsert" or "delete".
+        :return: Retuns the path to the transformed file.
+        """
+
+        assert self.datafile_release is not None, "Datafile.transform_file_path: self.datafile_release is None"
+
+        # Remove the .gz compression if it's not a baseline file.
+        if self.baseline:
+            file_type = self.file_type
+        else:
+            file_type = self.file_type[:-3]
+
+        return os.path.join(
+            self.datafile_release.transform_folder,
+            f"{entity_type}_{self.filename[:-7]}.{file_type}",
+        )
```

Comment:
It might be easier to see what is going on by using @property and just hard coding different bits of the file paths.

```python

@property
def transform_baseline_path():
    return os.path.join(self.datafile_release.transform_folder, f"baseline_{self.filename[:-7]}.jsonl.gz")

@property
def transform_upsert_path():
    return os.path.join(self.datafile_release.transform_folder, f"upsert_{self.filename[:-7]}.jsonl")

@property
def merged_upsert_path():
    return os.path.join(self.datafile_release.transform_folder, f"upsert_merged_{self.filename[:-7]}.jsonl.gz")

@property
def delete_file_path():
    return os.path.join(self.datafile_release.transform_folder, f"delete.jsonl.gz")

```